samples/matter/common/src/util/finite_map.h

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <functional>
+#include <map>
+
+namespace Nrf
+{
+/*
+   FiniteMap template container allows to wrap mappings between T1-type trivial key and T2-type value.
+   The size or the container is predefined at compilation time, therefore
+   the maximum number of stored elements must be well known. FiniteMap owns
+   inserted values, meaning that once the user inserts a value it will not longer
+   be valid outside of the container, i.e. FiniteMap cannot return stored object by copy.
+   FiniteMap's API offers basic operations like:
+     * inserting new values under provided key (Insert)
+     * erasing existing items under a given key (Erase) - this operation does not take care of the
+	   duplicated items,  meaning that the pointer under a given key is released unconditionally
+	   and the responsibility for the duplications is on the application side.
+     * retrieving values stored under provided key ([] operator)
+     * checking if the map contains a non-null value under given key (Contains)
+	 * retrieving a number of free slots available in the map (FreeSlots)
+     * iterating though stored item via publicly available mMap member
+     * retrieving the first free slot (GetFirstFreeSlot)
+	Prerequisites:
+     * T1 must be trivial.
+     * T2 must have move semantics and bool()/==operators implemented
+*/
+template <typename T1, typename T2, std::size_t N> struct FiniteMap {
```

Comment:
The problem with this template parameter T1 is that it must be unsigned numeric type, because it's used for accessing array elements. So, I suggest adding the following assert, or something similar:

`static_assert(std::is_integral_v<T1> && !std::numeric_limits<T1>::is_signed);`

 It'd be also good to mention this in the commentary above (T1 must be unsigned integer type).
 
 I think we should also at least do some smoke tests of the Matter Bridge application before merging this modification.

Reply:
Ok, I did not notice the implementation had been adapted for storing Items on the first free slot instead of the index corresponding to the Item.key. So in theory there are no such limitations on the T1 but we need to select the `kInvalidKey` value differently (as now we can store keys of values outside of the `mMap` range) - maybe the maximum of the numeric limits of T1 would be ok (let's then add this information to the map description).